core-backend/wasmi/src/env.rs

```diff
@@ -0,0 +1,502 @@
+                    reason: WasmiEnvironmentError::CreateEnvMemory(e),
+                    gas_amount: ext_carrier.into_inner().into_gas_amount(),
+                })
+            }
+        };
+
+        builder.add_memory("env", "memory", mem.clone());
+        builder.add_host_func("env", "alloc", Funcs::alloc);
+        builder.add_host_func("env", "free", Funcs::free);
+        builder.add_host_func("env", "gas", Funcs::gas);
+
+        let runtime = Runtime {
+            ext: ext_carrier,
+            memory: MemoryWrap::new(mem),
+            err: FuncError::Terminated(TerminationReason::Success),
+        };
+
+        let defined_host_functions = builder.defined_host_functions.clone();
+        let instance = match ModuleInstance::new(
+            &wasmi::Module::from_buffer(binary).expect("wasmi can't load module binary"),
+            &builder,
+        ) {
+            Ok(inst) => inst.not_started_instance().clone(),
+            Err(e) => {
+                return Err(BackendError {
+                    reason: WasmiEnvironmentError::ModuleInstantiation(e),
+                    gas_amount: runtime.ext.into_inner().into_gas_amount(),
+                })
+            }
+        };
+
+        Ok(WasmiEnvironment {
+            runtime,
+            instance,
+            defined_host_functions,
+            entries,
+        })
+    }
+
+    fn get_stack_mem_end(&mut self) -> Option<WasmPageNumber> {
+        // '__gear_stack_end' export is inserted in wasm-proc or wasm-builder
+        let global = self
+            .instance
+            .export_by_name("__gear_stack_end")?
+            .as_global()?
+            .get();
+        global.try_into::<i32>().and_then(|addr| {
+            if addr < 0 {
+                None
+            } else {
+                Some(WasmPageNumber(
+                    (addr as usize / WasmPageNumber::size()) as u32,
+                ))
+            }
+        })
+    }
+
+    fn get_mem(&self) -> &Self::Memory {
+        &self.runtime.memory
+    }
+
+    fn get_mem_mut(&mut self) -> &mut Self::Memory {
+        &mut self.runtime.memory
+    }
+
+    fn execute<F, T>(
+        mut self,
+        entry_point: &DispatchKind,
+        post_execution_handler: F,
+    ) -> Result<BackendReport, BackendError<Self::Error>>
+    where
+        F: FnOnce(&Self::Memory) -> Result<(), T>,
+        T: fmt::Display,
+    {
+        let res = if self.entries.contains(entry_point) {
+            let mut externals = GuestExternals {
+                state: &mut self.runtime,
+                defined_host_functions: &self.defined_host_functions,
+            };
+            self.instance
+                .invoke_export(entry_point.into_entry(), &[], &mut externals)
+                .map(|_| ())
+        } else {
+            Ok(())
+        };
+
+        let Runtime {
+            ext,
+            memory,
+            err: trap,
+        } = self.runtime;
+
+        log::debug!("execution res = {:?}", res);
```

Comment:
I would add more details to the log:
```suggestion
        log::debug!("WasmiEnvironment::execute result = {res:?}");
```